lib/synapse/retry_policy.rb

```diff
@@ -0,0 +1,37 @@
+module Synapse
+  module RetryPolicy
+    def with_retry(options = {}, &callback)
+      max_attempts = options[:max_attempts] || 1
+      base_interval = options[:base_interval] || 0
+      max_interval = options[:max_interval] || 0
+      retriable_errors = Array(options[:retriable_errors] || StandardError)
+      if max_attempts <= 0
+        raise ArgumentError, ":max_attempts must be greater than 0"
+      end
+      if base_interval > max_interval
+        raise ArgumentError, ":base_interval cannot be greater than :max_interval."
+      end
+      if callback.nil?
+        raise ArgumentError, "callback cannot be nil"
+      end
+
+      attempts = 0
+      start_time = Time.now
+      begin
+        attempts += 1
+        return callback.call(attempts)
+      rescue *retriable_errors => error
+        if attempts >= max_attempts
+          raise error
+        end
+        sleep get_retry_interval(base_interval, max_interval, attempts, Time.now - start_time)
+        retry
+      end
+    end
+
+    def get_retry_interval(base_interval, max_interval, attempts, elapsed)
```

Comment:
What is the point in subtracting out `elapsed` here?

It adds in additional complexity (in terms of reasoning about how long the `n`th attempt will wait for). In addition, I don't think that's standard in typical exponential back-off algorithms.

Reply:
If we don't consider how long it takes a call to fail, the interval btw retries can be skewed. Say we want to delay 60 seconds btw retires. A call failed with connection lost can only take 1 second, the retry interval becomes 61 seconds. A call failed with operation timeout can take 30 seconds, the retry interval becomes 90 seconds. As retry interval accumulates, the total delay of API call can be skewed significantly.

Agree this is not standard. But it is quite common variant to take actual call delay into consideration:
https://github.com/cenkalti/backoff/blob/2146c93394225c3732078705043ce9f26584d334/exponential.go#L128

https://github.com/apache/incubator-gobblin/blob/master/gobblin-utility/src/main/java/org/apache/gobblin/util/ExponentialBackoff.java#L72

As we care about the upper bound delay of retry, we can expose another parameter called `max_delay` to stop retry once total elapsed reached the limit.